morango/api/viewsets.py

```diff
@@ -0,0 +1,273 @@
+        local_scope = syncsession.local_certificate.get_scope()
+        if request.data.get("push"):
+            if not requested_filter.is_subset_of(remote_scope.write_filter):
+                scope_error_msg = "Client certificate scope does not permit pushing for the requested filter."
+            if not requested_filter.is_subset_of(local_scope.read_filter):
+                scope_error_msg = "Server certificate scope does not permit receiving pushes for the requested filter."
+        else:
+            if not requested_filter.is_subset_of(remote_scope.read_filter):
+                scope_error_msg = "Client certificate scope does not permit pulling for the requested filter."
+            if not requested_filter.is_subset_of(local_scope.write_filter):
+                scope_error_msg = "Server certificate scope does not permit responding to pulls for the requested filter."
+        if scope_error_msg:
+            return response.Response(
+                scope_error_msg,
+                status=status.HTTP_403_FORBIDDEN
+            )
+
+        # build the data to be used for creating the transfersession
+        data = {
+            "id": request.data.get("id"),
+            "start_timestamp": timezone.now(),
+            "last_activity_timestamp": timezone.now(),
+            "active": True,
+            "filter": request.data.get("filter"),
+            "incoming": request.data.get("push"),
+            "records_total": request.data.get("records_total") if is_a_push else None,
+            "sync_session": syncsession,
+        }
+
+        transfersession = models.TransferSession(**data)
+        transfersession.full_clean()
+        transfersession.save()
+
+        return response.Response(
+            serializers.TransferSessionSerializer(transfersession).data,
+            status=status.HTTP_201_CREATED,
+        )
+
+    def perform_destroy(self, transfersession):
+        transfersession.active = False
+        transfersession.save()
+
+    def get_queryset(self):
+        return models.TransferSession.objects.filter(active=True)
+
+
+class BufferViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    permission_classes = (permissions.BufferPermissions,)
+    serializer_class = serializers.BufferSerializer
+    pagination_class = pagination.LimitOffsetPagination
+
+    def create(self, request):
+        data = request.data if isinstance(request.data, list) else [request.data]
+        serial_data = serializers.BufferSerializer(data=data, many=True)
+        if serial_data.is_valid():
+
+            # ensure the transfer session allows pushes, and is same across records
+            if not serial_data.validated_data[0]["transfer_session"].incoming:
+                return response.Response(
+                    "Specified TransferSession does not allow pushes.",
+                    status=status.HTTP_403_FORBIDDEN
+                )
+            if len(set(rec["transfer_session"] for rec in serial_data.data)) > 1:
```

Comment:
In part it's a sanity check, but it's also for security reasons, as in the check on line 251 above we only check against the first record's transfer_session.